### PR TITLE
add syntax for empty n-dimensional arrays

### LINKED ISSUE #41389: syntax for creating empty n-dimensional arrays
These currently throw an error:

```julia
julia> [;]
ERROR: syntax: unexpected ";"
Stacktrace:
 [1] top-level scope
   @ none:1

julia> [;;]
ERROR: syntax: unexpected ";"
Stacktrace:
 [1] top-level scope
   @ none:1

julia> [;;;]
ERROR: syntax: unexpected ";"
Stacktrace:
 [1] top-level scope
   @ none:1
```

I think it would make sense for them to return an empty vector/0×0 matrix/0×0×0 array respectively.

### PR DESCRIPTION
**Pull request by simeonschaub**
Might be worth backporting to 1.7 for consistency? Otherwise, this should probably have a NEWS entry.

closes #41389
cc: @BioTurboNick

### CODE REVIEW

**Comment by BioTurboNick** [comment]
Nice, you didn't even have to touch `julia-syntax.scm` to make it work.

Currently the parser allows this:
```
[;;
;]
0×0×0 Array{Any, 3}
```

Line breaks between semicolons produce an error normally:
```
[1 ;;
;]
ERROR: syntax: semicolons may appear before or after a line break in an array expression, but not both
Stacktrace:
 [1] top-level scope
   @ none:1
```

**Comment by simeonschaub** [comment]
> Currently the parser allows this:
> 
> ```
> [;;
> ;]
> 0×0×0 Array{Any, 3}
> ```

Oh, weird, I thought the `disallow-space` would catch this.

File: test/syntax.jl [hunk 1]
```diff
@@ -2959,3 +2959,30 @@ end
...
+    @test :(T[;])   == Expr(:typed_ncat, :T, 1)
+    @test :(T[;;])  == Expr(:typed_ncat, :T, 2)
+    @test :(T[;;;]) == Expr(:typed_ncat, :T, 3)
+
+    @test Int[]    == Array{Int}(undef, 0)
+    @test Int[;]   == Array{Int}(undef, 0)
+    @test Int[;;]  == Array{Int}(undef, 0, 0)
+    @test Int[;;;] == Array{Int}(undef, 0, 0, 0)
+
+    @test_throws ParseError Meta.parse("[ ;]")
```
Comment (StefanKarpinski) [inline test/syntax.jl:2984]: I'm not sure that leading/trailing spaces should be an error...

Comment (simeonschaub) [reply test/syntax.jl:2984]: Yeah, I think I agree.

**Comment by vtjnash** [comment]
bump?

**Comment by PallHaraldsson** [comment]
Why probably not better to backport? If 1.7 ends up as LTS, then it's very useful to have all new syntax in. [I believe syntax can't be added with Compat.jl.]

**Comment by simeonschaub** [comment]
1.7 won't become the LTS. The same argument could be made for any feature. 1.7 is already taking longer to finalize than usual, so I don't think we should stretch that out any further by backporting additional feature.

**Pull request merged**